File: src/com/lapissea/vulkanimpl/util/types/VkDescriptorPool.java
package com.lapissea.vulkanimpl.util.types;

import com.lapissea.vulkanimpl.Vk;
import com.lapissea.vulkanimpl.VkGpu;
import com.lapissea.vulkanimpl.util.VkConstruct;
import com.lapissea.vulkanimpl.util.VkDestroyable;
import com.lapissea.vulkanimpl.util.VkGpuCtx;
import org.lwjgl.vulkan.VkDescriptorSetAllocateInfo;

import java.nio.LongBuffer;

import static org.lwjgl.system.MemoryUtil.*;
import static org.lwjgl.vulkan.VK10.*;

public class VkDescriptorPool implements VkDestroyable, VkGpuCtx{
	private final VkGpu gpu;
	private final long  handle;
	
	public VkDescriptorPool(VkGpu gpu, long handle){
		this.gpu=gpu;
		this.handle=handle;
	}
	
	public long allocateDescriptorSets(LongBuffer layouts){
		LongBuffer dest=memAllocLong(1);
		try(VkDescriptorSetAllocateInfo allocInfo=VkConstruct.descriptorSetAllocateInfo()){
			allocInfo.descriptorPool(handle)
			         .pSetLayouts(layouts);
			
			return Vk.allocateDescriptorSets(gpu, allocInfo, dest);
		}finally{
			memFree(dest);
		}
	}
	
	public long getHandle(){
		return handle;
	}
	
	@Override
	public void destroy(){
		vkDestroyDescriptorPool(getDevice(), handle, null);
	}
	
	@Override
	public VkGpu getGpu(){
		return gpu;
	}
}

File: src/com/lapissea/vulkanimpl/util/types/VkDescriptorPool.kt
package com.lapissea.vulkanimpl.util.types

import com.lapissea.vulkanimpl.Vk
import com.lapissea.vulkanimpl.VkGpu
import com.lapissea.vulkanimpl.util.VkConstruct
import com.lapissea.vulkanimpl.util.VkDestroyable
import com.lapissea.vulkanimpl.util.VkGpuCtx
import org.lwjgl.vulkan.VK10.*
import java.nio.LongBuffer

class VkDescriptorPool(private val gpu: VkGpu, val handle: Long): VkDestroyable, VkGpuCtx {
	
	fun allocateDescriptorSets(layouts: VkDescriptorSetLayout)=allocateDescriptorSets(layouts.buff)[0]
	
	fun allocateDescriptorSets(layouts: LongBuffer): Array<VkDescriptorSet> {
		VkConstruct.descriptorSetAllocateInfo().use {
			
			it.descriptorPool(handle).pSetLayouts(layouts)
			
			return Vk.allocateDescriptorSets(gpu, it)
		}
	}
	
	override fun destroy()=vkDestroyDescriptorPool(device, handle, null)
	
	override fun getGpu()=gpu
	
}
